### PR TITLE
[feature-wip] (array-type) add the array_distinct function

### LINKED ISSUE #10052: [Feature] Add lots of Array Functions
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

We have supported `array` data type recently, all pr(s) about `array` please refer to https://github.com/apache/incubator-doris/issues/7570 .

Now we need to add lots of array functions to make `array` more powerful.
If anybody is interesting to `array`, please let me know.

Array Functions DONE:
- size/cardinality
- array_position
- array_contains
- element_at

Array Functions DOING:
- array_sum
- array_avg
- array_min
- array_max
- array_product
- array_distinct
- array_remove
- arrays_overlap

Array Functions TODO:
- array_union
- array_intersect
- array_except
- reverse
- concat_ws
- array_join
- array_sort
- array_slice

### Use case

_No response_

### Related issues

https://github.com/apache/incubator-doris/issues/7570

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by carlvinhust2012**
# Proposed changes
Issue Number:close #10052 
1. add the  array_distinct function for array-type which follow spark syntax；
2. we can run this function in sql as follow:

MySQL [(none)]> set enable_array_type=true;

MySQL [(none)]> set enable_vectorized_engine=true;

MySQL [(none)]> select id, c_array, array_distinct(c_array) from example_db.array_test;
+------+-----------------------------+---------------------------+
| id   | c_array                     | array_distinct(`c_array`) |
+------+-----------------------------+---------------------------+
|    1 | [1, 2, 3, 4, 5]             | [1, 2, 3, 4, 5]           |
|    2 | [6, 7, 8]                   | [6, 7, 8]                 |
|    3 | []                          | []                        |
|    4 | NULL                        | NULL                      |
|    5 | [1, 2, 3, 4, 5, 4, 3, 2, 1] | [1, 2, 3, 4, 5]           |
|    6 | [1, 2, 3, NULL]             | [1, 2, 3, NULL]           |
|    7 | [1, 2, 3, NULL, NULL]       | [1, 2, 3, NULL, NULL]     |
+------+-----------------------------+---------------------------+
7 rows in set (0.005 sec)

MySQL [(none)]> select id, c_array, array_distinct(c_array) from example_db.array_test01;
+------+--------------------------+---------------------------+
| id   | c_array                  | array_distinct(`c_array`) |
+------+--------------------------+---------------------------+
|    1 | [a, b, c, d, e]          | [a, b, c, d, e]           |
|    2 | [f, g, h]                | [f, g, h]                 |
|    3 | []                       | []                        |
|    3 | [NULL]                   | [NULL]                    |
|    5 | [a, b, c, d, e, a, b, c] | [a, b, c, d, e]           |
|    6 | NULL                     | NULL                      |
|    7 | [a, b, NULL]             | [a, b, NULL]              |
|    7 | [a, b, NULL, NULL]       | [a, b, NULL, NULL]        |
+------+--------------------------+---------------------------+
8 rows in set (0.009 sec)

3. the 'BufferWritable'  was defined in 'vec/common/string_buffer.hpp', so  'io_helper.h' should include this header file.

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (No)
5. Has unit tests been added: (No)
6. Has document been added or modified: (No)
7. Does it need to update dependencies: (No)
8. Are there any changes that cannot be rolled back: (No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/functions/array/function_array_distinct.h [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+    /// Get function name.
+    String get_name() const override { return name; }
+
+    bool is_variadic() const override { return false; }
+
+    size_t get_number_of_arguments() const override { return 1; }
+
+    DataTypePtr get_return_type_impl(const DataTypes& arguments) const override {
+        DCHECK(is_array(arguments[0]))
+                << "first argument for function: " << name << " should be DataTypeArray";
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_distinct.h:49]: print arguments[0] while DCHECK failed.

File: be/src/vec/functions/array/function_array_distinct.h [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+        return make_nullable(
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_distinct.h:50]: output datatype should be the same as input

File: be/src/vec/functions/array/function_array_distinct.h [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+            return Status::RuntimeError(
+                    fmt::format("unsupported types for function {}({})", get_name(),
+                                block.get_by_position(arguments[0]).type->get_name()));
+        }
+        const auto& src_offsets = src_column_array->get_offsets();
+        const auto& src_column_data = src_column_array->get_data();
+
+        DataTypePtr src_column_type = remove_nullable(block.get_by_position(arguments[0]).type);
+        auto nested_type = assert_cast<const DataTypeArray&>(*src_column_type).get_nested_type();
+        auto dest_column_ptr = ColumnArray::create(nested_type->create_column(),
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_distinct.h:69]: clone from src column is enough?

File: be/src/vec/functions/array/function_array_distinct.h [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+                                                   ColumnArray::ColumnOffsets::create());
+        IColumn& dest_column = dest_column_ptr->get_data();
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_distinct.h:71]: dest_nested_column?

File: be/src/vec/functions/array/function_array_distinct.h [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+        ColumnArray::Offsets& dest_offsets = dest_column_ptr->get_offsets();
+
+        const IColumn* src_inner_column = nullptr;
+        const ColumnNullable* src_nullable_column =
+                check_and_get_column<ColumnNullable>(src_column_data);
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_distinct.h:76]: normally we define variables like:
xxx_data ===> pod_array
xxx_column or xxx_col  ===> column

no need to define src_column_data.

File: be/src/vec/functions/array/function_array_distinct.h [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+
+        DataTypePtr src_column_type = remove_nullable(block.get_by_position(arguments[0]).type);
+        auto nested_type = assert_cast<const DataTypeArray&>(*src_column_type).get_nested_type();
+        auto dest_column_ptr = ColumnArray::create(nested_type->create_column(),
+                                                   ColumnArray::ColumnOffsets::create());
+        IColumn& dest_column = dest_column_ptr->get_data();
+        ColumnArray::Offsets& dest_offsets = dest_column_ptr->get_offsets();
+
+        const IColumn* src_inner_column = nullptr;
+        const ColumnNullable* src_nullable_column =
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_distinct.h:75]: src_nested_nullable_col

File: be/src/vec/functions/array/function_array_distinct.h [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+    bool _execute_by_type(const IColumn& src_column, const ColumnArray::Offsets& src_offsets,
+                          IColumn& dest_column, ColumnArray::Offsets& dest_offsets,
+                          const ColumnNullable* src_nullable_col, ColumnNullable* dest_nullable_col,
+                          DataTypePtr& nested_type) {
+        bool res = false;
+        WhichDataType which(remove_nullable(nested_type)->get_type_id());
+        if (which.idx == TypeIndex::UInt8) {
+            res = _execute_number<UInt8>(src_column, src_offsets, dest_column, dest_offsets,
+                                         src_nullable_col, dest_nullable_col);
+        } else if (which.idx == TypeIndex::UInt16) {
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_distinct.h:247]: we do not support Unsigned data type, except UInt8 for Boolean.

File: be/src/vec/functions/array/function_array_distinct.h [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+        } else if (which.idx == TypeIndex::Float32) {
+            res = _execute_number<Float32>(src_column, src_offsets, dest_column, dest_offsets,
+                                           src_nullable_col, dest_nullable_col);
+        } else if (which.idx == TypeIndex::Float64) {
+            res = _execute_number<Float64>(src_column, src_offsets, dest_column, dest_offsets,
+                                           src_nullable_col, dest_nullable_col);
+        } else if (which.idx == TypeIndex::Date) {
+            res = _execute_number<Date>(src_column, src_offsets, dest_column, dest_offsets,
+                                        src_nullable_col, dest_nullable_col);
+        } else if (which.idx == TypeIndex::DateTime) {
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_distinct.h:283]: need add decimal support

File: be/src/vec/functions/array/function_array_distinct.h [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+            return false;
+        }
+        const PaddedPODArray<T>& src_datas = src_data_concrete->get_data();
+
+        ColumnVector<T>& dest_data_concrete = reinterpret_cast<ColumnVector<T>&>(dest_column);
+        PaddedPODArray<T>& dest_datas = dest_data_concrete.get_data();
+
+        const PaddedPODArray<UInt8>* src_null_map = nullptr;
+        if (src_nullable_col) {
+            src_null_map = &src_nullable_col->get_null_map_column().get_data();
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_distinct.h:125]: src_nullable_col->get_null_map_column().get_data().data() returns UInt8*.

use `const UInt8* src_null_map` is simpler.

File: be/src/vec/functions/array/function_array_distinct.h (hunk 2 above)
Comment (carlvinhust2012) [reply be/src/vec/functions/array/function_array_distinct.h:50]: > 

ok，here not need add make_nullable.

File: be/src/vec/functions/array/function_array_distinct.h (hunk 4 above)
Comment (carlvinhust2012) [reply be/src/vec/functions/array/function_array_distinct.h:71]: > 

yes, it also can use dest_nested_column.

File: be/src/vec/functions/array/function_array_distinct.h (hunk 9 above)
Comment (carlvinhust2012) [reply be/src/vec/functions/array/function_array_distinct.h:125]: > src_nullable_col->get_null_map_column().get_data().data() returns UInt8*.
> 
> use `const UInt8* src_null_map` is simpler.

this is a good suggestion,  I will simplify here.

File: be/src/vec/functions/array/function_array_distinct.h (hunk 7 above)
Comment (carlvinhust2012) [reply be/src/vec/functions/array/function_array_distinct.h:247]: > we do not support Unsigned data type, except UInt8 for Boolean.

ok, remove it later.

File: be/src/vec/functions/array/function_array_distinct.h (hunk 8 above)
Comment (carlvinhust2012) [reply be/src/vec/functions/array/function_array_distinct.h:283]: > need add decimal support

yes, I am adding function for the decimal.

File: gensrc/script/doris_builtins_functions.py [hunk 10]
```diff
@@ -160,6 +160,18 @@
     [['array_position'], 'BIGINT', ['ARRAY_STRING', 'STRING'], '', '', '', 'vec', ''],
 
     [['cardinality', 'size'], 'BIGINT', ['ARRAY'], '', '', '', 'vec', ''],
+    [['array_distinct'], 'ARRAY_TINYINT',   ['ARRAY_TINYINT'], '', '', '', 'vec', ''],
```
Comment (eldenmoon) [inline gensrc/script/doris_builtins_functions.py:163]: why not add symbol to functions

Comment (cambyzju) [reply gensrc/script/doris_builtins_functions.py:163]: Array functions only supported on vec engine. We do not need to add them in vec engine.

Comment (eldenmoon) [reply gensrc/script/doris_builtins_functions.py:163]: But the sql may fail if not add symbol, you could try this, here's my issue i encounterd

Comment (cambyzju) [reply gensrc/script/doris_builtins_functions.py:163]: Got it, I will find the problem and try to fix it. 
Some weeks before, where condition with array function works well.

Comment (cambyzju) [reply gensrc/script/doris_builtins_functions.py:163]: set enable_array_type = true;

Comment (eldenmoon) [reply gensrc/script/doris_builtins_functions.py:163]: sorry, it's still not working for me, errors still occur if you adding array function in `where` condition. eg. 
```
mysql> create table test_array_string (k1 INT, k2 INT, k3 array<string>) ENGINE=olap DUPLICATE KEY(k1, k2) PARTITION BY RANGE (k1) (partition `p1` values less than ("1000"), partition `p2` values less than ("2000"),partition `p3` values less than ("3000"))  DISTRIBUTED BY HASH(k2) BUCKETS 3 PROPERTIES("replication_num" = "1");
Query OK, 0 rows affected (0.02 sec)

mysql> insert into test_array_string  values(1, 2, ["a", "b", "c"]),(1, 2, ["a", "b"]),(1, 2, ["a", "xxqwdqw", "c"]),(1, 2, ["a", "b"]),(1, 2, ["a", "b", "c"]),(1, 2, ["a", "b", "c"]),(1, 2, ["a", "b", "c"]),(1, 2, ["a", "b"]),(1, 2, ["a", "b", "c", "d", "e"]),(1, 2, ["a", "b", "c"]), (1, 2, ["a", "b", "cdsdasd"]),(1, 2, [ "nishidabenzhu123"]),(1, 2, ["a", "b", "c"]),(1, 2, ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j"]), (1, 2, ["a", "cdsdasd"]), (1, 2, ["cdsdasd", "lihngyu2", "3", "4", "5"]);
Query OK, 16 rows affected (0.04 sec)
{'label':'insert_c3ab610af58b4b3a-83d3cc32cf9f1709', 'status':'VISIBLE', 'txnId':'4'}

mysql> set enable_vectorized_engine = true;
Query OK, 0 rows affected (0.01 sec)

mysql> set enable_array_type = true;
Query OK, 0 rows affected (0.00 sec)

mysql> select size(k3) from test_array_string where size(k3) != 3;
**ERROR 1105 (HY000): errCode = 2, detailMessage = Function size is not implemented.**

mysql> select size(k3) from test_array_string;
+------------+
| size(`k3`) |
+------------+
|          5 |
|          2 |
|         10 |
|          3 |
|          1 |
|          3 |
|          3 |
|          5 |
|          2 |
|          3 |
|          3 |
|          3 |
|          2 |
|          3 |
|          2 |
|          3 |
+------------+
16 rows in set (0.01 sec)
```

Comment (cambyzju) [reply gensrc/script/doris_builtins_functions.py:163]: Got it, I will try to reproduce and fix it.

Comment (HappenLee) [reply gensrc/script/doris_builtins_functions.py:163]: @eldenmoon hello, i had fix the bug in this pr: https://github.com/apache/doris/pull/10467

Comment (eldenmoon) [reply gensrc/script/doris_builtins_functions.py:163]: @HappenLee thanks a lot, i will test it immediately after it's merged

File: be/src/vec/functions/array/function_array_distinct.h [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+        const auto& src_nested_column = src_column_array->get_data();
+
+        DataTypePtr src_column_type = remove_nullable(block.get_by_position(arguments[0]).type);
+        auto nested_type = assert_cast<const DataTypeArray&>(*src_column_type).get_nested_type();
+        auto dest_column_ptr = ColumnArray::create(nested_type->create_column(),
+                                                   ColumnArray::ColumnOffsets::create());
+        IColumn& dest_nested_column = dest_column_ptr->get_data();
+        ColumnArray::Offsets& dest_offsets = dest_column_ptr->get_offsets();
+
+        const IColumn* src_inner_column = nullptr;
```
Comment (xy720) [inline be/src/vec/functions/array/function_array_distinct.h:74]: Use ColumnPtr instead is better.

File: be/src/vec/functions/array/function_array_distinct.h [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+        const ColumnNullable* src_nested_nullable_col = nullptr;
+        if (src_nested_column.is_nullable()) {
+            src_nested_nullable_col = check_and_get_column<ColumnNullable>(src_nested_column);
+            src_inner_column = src_nested_nullable_col->get_nested_column_ptr();
+        } else {
+            src_inner_column = src_column_array->get_data_ptr();
+        }
+
+        IColumn* dest_inner_column = nullptr;
```
Comment (xy720) [inline be/src/vec/functions/array/function_array_distinct.h:83]: Same above.

File: be/src/vec/functions/array/function_array_distinct.h [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+        }
+        const auto& src_offsets = src_column_array->get_offsets();
+        const auto& src_nested_column = src_column_array->get_data();
+
+        DataTypePtr src_column_type = remove_nullable(block.get_by_position(arguments[0]).type);
+        auto nested_type = assert_cast<const DataTypeArray&>(*src_column_type).get_nested_type();
+        auto dest_column_ptr = ColumnArray::create(nested_type->create_column(),
+                                                   ColumnArray::ColumnOffsets::create());
+        IColumn& dest_nested_column = dest_column_ptr->get_data();
+        ColumnArray::Offsets& dest_offsets = dest_column_ptr->get_offsets();
```
Comment (xy720) [inline be/src/vec/functions/array/function_array_distinct.h:72]: The size of dest_offsets should be same as input_rows_count, 
so I think you can reserve it like `dest_offsets.reserve(input_rows_count)`

Comment (carlvinhust2012) [reply be/src/vec/functions/array/function_array_distinct.h:72]: yes, it is a good suggestion. I will add this.

File: be/src/vec/functions/array/function_array_distinct.h [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+    String get_name() const override { return name; }
+
+    bool is_variadic() const override { return false; }
+
+    size_t get_number_of_arguments() const override { return 1; }
+
+    DataTypePtr get_return_type_impl(const DataTypes& arguments) const override {
+        DCHECK(is_array(arguments[0]))
+                << "first argument for function: " << name << " should be DataTypeArray"
+                << " and arguments[0] is " << arguments[0].get()->get_name();
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_distinct.h:50]:
```suggestion
                << " and arguments[0] is " << arguments[0]->get_name();
```

File: be/src/vec/functions/array/function_array_distinct.h [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+        return check_and_get_data_type<DataTypeArray>(arguments[0].get())->get_nested_type();
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_distinct.h:51]: why not return type is not arguments[0]？

Comment (carlvinhust2012) [reply be/src/vec/functions/array/function_array_distinct.h:51]: this is a problem, so return arguments[0] is more better.

File: be/src/vec/functions/array/function_array_distinct.h (hunk 11 above)
Comment (carlvinhust2012) [reply be/src/vec/functions/array/function_array_distinct.h:75]: I think that it use IColumn directly can reduce type conversions.

File: be/src/vec/functions/array/function_array_distinct.h [hunk 16]
```diff
@@ -0,0 +1,297 @@
...
+
+        block.replace_by_position(result, std::move(dest_column_ptr));
+        return Status::OK();
+    }
+
+private:
+    // Note: Here initially allocate a piece of memory for 2^5 = 32 elements.
+    static constexpr size_t INITIAL_SIZE_DEGREE = 5;
+
+    template <typename ColumnType, typename NestType>
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_distinct.h:111]: ColumnType or NestType, only one is needed:
Input ColumnType, using NestType = ColumnType::value_type
Input NestType, using ColumnType = ColumnVector<NestType>

File: be/src/vec/functions/array/function_array_distinct.h [hunk 17]
```diff
@@ -0,0 +1,297 @@
...
+        if (src_nullable_col) {
+            src_null_map = &src_nullable_col->get_null_map_column().get_data();
+        }
+
+        PaddedPODArray<UInt8>* dest_null_map = nullptr;
+        if (dest_nullable_col) {
+            dest_null_map = &dest_nullable_col->get_null_map_column().get_data();
+        }
+
+        using Set = HashSetWithSavedHashWithStackMemory<UInt128, UInt128TrivialHash,
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_distinct.h:135]: UInt128 for all type?

File: be/src/vec/functions/array/function_array_distinct.h [hunk 18]
```diff
@@ -0,0 +1,265 @@
...
+        const auto& src_column_array = check_and_get_column<ColumnArray>(*src_column);
+        if (!src_column_array) {
+            return Status::RuntimeError(
+                    fmt::format("unsupported types for function {}({})", get_name(),
+                                block.get_by_position(arguments[0]).type->get_name()));
+        }
+        const auto& src_offsets = src_column_array->get_offsets();
+        const auto& src_nested_column = src_column_array->get_data();
+
+        DataTypePtr src_column_type = remove_nullable(block.get_by_position(arguments[0]).type);
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_distinct.h:68]: we use default nullable process logic, input type will not be nullable, no need to remove_nullable.

Comment (carlvinhust2012) [reply be/src/vec/functions/array/function_array_distinct.h:68]: ok.I will re-check it. Maybe add 'remove_nullable' could be protective?

File: be/src/vec/functions/array/function_array_distinct.h [hunk 19]
```diff
@@ -0,0 +1,268 @@
...
+        }
+        return true;
+    }
+
+    bool _execute_by_type(const IColumn& src_column, const ColumnArray::Offsets& src_offsets,
+                          IColumn& dest_column, ColumnArray::Offsets& dest_offsets,
+                          const NullMapType* src_null_map, NullMapType* dest_null_map,
+                          DataTypePtr& nested_type) {
+        bool res = false;
+        WhichDataType which(remove_nullable(nested_type)->get_type_id());
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_distinct.h:226]:
```suggestion
        WhichDataType which(remove_nullable(nested_type));
```

File: be/src/vec/functions/array/function_array_distinct.h [hunk 20]
```diff
@@ -0,0 +1,268 @@
...
+        if (which.idx == TypeIndex::UInt8) {
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_distinct.h:227]:
```suggestion
        if (which.is_uint8()) {
```

File: be/src/vec/functions/array/function_array_distinct.h [hunk 21]
```diff
@@ -0,0 +1,268 @@
...
+    static constexpr size_t INITIAL_SIZE_DEGREE = 5;
+
+    template <typename ColumnType>
+    bool _execute_number(const IColumn& src_column, const ColumnArray::Offsets& src_offsets,
+                         IColumn& dest_column, ColumnArray::Offsets& dest_offsets,
+                         const NullMapType* src_null_map, NullMapType* dest_null_map) {
+        using NestType = typename ColumnType::value_type;
+        using ElementNativeType = typename NativeType<NestType>::Type;
+
+        const ColumnType* src_data_concrete = check_and_get_column<ColumnType>(&src_column);
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_distinct.h:122]: use reinterpret_cast

File: be/src/vec/functions/array/function_array_distinct.h [hunk 22]
```diff
@@ -0,0 +1,268 @@
...
+        Set set;
+
+        ColumnArray::Offset prev_src_offset = 0;
+        ColumnArray::Offset res_offset = 0;
+
+        for (auto curr_src_offset : src_offsets) {
+            set.clear();
+            size_t null_size = 0;
+            for (ColumnArray::Offset j = prev_src_offset; j < curr_src_offset; ++j) {
+                if ((*src_null_map)[j]) {
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_distinct.h:142]: check src_null_map is nullable

File: be/src/vec/functions/array/function_array_distinct.h [hunk 23]
```diff
@@ -0,0 +1,268 @@
...
+                    if (dest_null_map) {
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_distinct.h:143]: assert or DCHECK dest_null_map

File: be/src/vec/functions/array/function_array_distinct.h [hunk 24]
```diff
@@ -0,0 +1,268 @@
...
+        const ColumnString* src_data_concrete = check_and_get_column<ColumnString>(&src_column);
+        if (!src_data_concrete) {
+            return false;
+        }
+
+        ColumnString& dest_column_string = reinterpret_cast<ColumnString&>(dest_column);
+        ColumnString::Chars& column_string_chars = dest_column_string.get_chars();
+        ColumnString::Offsets& column_string_offsets = dest_column_string.get_offsets();
+        column_string_chars.reserve(src_column.size());
+        column_string_offsets.reserve(src_offsets.size());
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_distinct.h:182]: already reserve ColumnString

File: docs/en/docs/sql-manual/sql-functions/array-functions/array_distinct.md [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+| 5    | [a, b, c, d, e, a, b, c] | [a, b, c, d, e]           |
+| 6    | NULL                     | NULL                      |
+| 7    | [a, b, NULL]             | [a, b, NULL]              |
+| 8    | [a, b, NULL, NULL]       | [a, b, NULL, NULL]        |
++------+--------------------------+---------------------------+
+```
+
+### keywords
+
+ARRAY, DISTINCT
```
Comment (cambyzju) [inline docs/en/docs/sql-manual/sql-functions/array-functions/array_distinct.md:79]: ARRAY_DISTINCT

**Comment by cambyzju** [comment]
LGTM

**Review by yangzhg — APPROVED** [review]
LGTM

**Pull request merged**